pkg/s3store/s3store.go

```diff
@@ -100,6 +101,91 @@ import (
 // considered valid into a header value according to RFC2616.
 var nonPrintableRegexp = regexp.MustCompile(`[^\x09\x20-\x7E]`)
 
+var _ handler.ServerDataStore = &S3Store{}
+
+func (store S3Store) AsServableUpload(upload handler.Upload) handler.ServableUpload {
+	return &S3ServableUpload{
+		store:  &store,
+		upload: upload.(*s3Upload),
+	}
+}
+
+type S3ServableUpload struct {
+	store  *S3Store
+	upload *s3Upload
+}
+
+func (su *S3ServableUpload) ServeContent(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	// Get file info
+	info, err := su.upload.GetInfo(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Prepare GetObject input
+	input := &s3.GetObjectInput{
+		Bucket: aws.String(su.store.Bucket),
+		Key:    su.store.keyWithPrefix(su.upload.objectId),
+	}
+
+	// Handle range requests
+	rangeHeader := r.Header.Get("Range")
+	if rangeHeader != "" {
+		if err := su.handleRangeRequest(ctx, w, r, info, input, rangeHeader); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// For non-range requests, serve the entire file
+	result, err := su.store.Service.GetObject(ctx, input)
+	if err != nil {
+		return err
+	}
+	defer result.Body.Close()
+
+	// Set headers
+	w.Header().Set("Content-Length", strconv.FormatInt(info.Size, 10))
+	w.Header().Set("Content-Type", info.MetaData["filetype"])
+	w.Header().Set("ETag", *result.ETag)
+
+	// Stream the content
+	_, err = su.store.copyToResponseWriter(w, result.Body)
+	return err
+}
+
+func (su *S3ServableUpload) handleRangeRequest(ctx context.Context, w http.ResponseWriter, r *http.Request, info handler.FileInfo, input *s3.GetObjectInput, rangeHeader string) error {
```

Comment:
I'm wonder whether dedicated handling for range requests is necessary. If the request includes the Range header we can add it to `GetObject` without parsing the ranges and let S3 deal with it. Afterwards we just copy the Content-Range, Content-Length and ETag header from S3's response to tusd's response and copy the body.

The advantage is that we don't have to parse the header (and possibly introduce subtle bugs) and don't have to construct the Content-Range and Content-Length headers on our own.

We would also not need a dedicated `handleRangeRequest` function as this could all live in `ServeContent`.

What do you think?

Reply:
I believe given that multiple ranges are possible, it made since to check it ourselves. and due to the complexity of the range checking, its why I had a unit test for the range logic made.

I can punt it all to be forwarded to s3, but I also think it is valid to verify at-least the ranges earlier in  the s3 impl.

Reply:
The goal of the new ServeContent interface is to reduce the complexity that we have to handle inside of tusd and adding logic to each store to validate the Range header is not having the opposite effect. In addition, deferring the validation to S3 (or whatever provider other stores use) is preferable since it knows best which values it will accept and we avoid discrepancies between tusd's validation and S3's.

Reply:
done